### ISSUE TITLE
Dashboard generates too many requests when listing resources

### ISSUE THREAD

**Issue by tkolo** [rancher/rancher#30718]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
1, Go to cluster explorer
2. Go to ingress page

**Result:**
Depending on the browser, it either completely hangs or works really slowly. It seems to go into infinite loop trying to fetch cluster's resoruces for some reason. 

![obraz](https://user-images.githubusercontent.com/3638714/103905267-b0058880-50fe-11eb-83b8-35b01a6c9059.png)

It did work correctly on "stable" branch which I believe is 2.4.x

**Other details that may be helpful:**
v2.5.3

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.5.3
- Installation option (single install/HA): HA, set up using RKE

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): local/imported, but created using RKE
- Machine type (cloud/VM/metal) and specifications (CPU/memory): metal
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"archive", BuildDate:"2020-11-25T13:19:56Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:41:49Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}

```

- Docker version (use `docker version`):

```
Client: Docker Engine - Community
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 17:02:52 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:01:20 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```

gz#15108

gz#17198

**Comment by axeal** [comment]
**What kind of request is this (question/bug/enhancement/feature request):**

Bug

**Steps to reproduce (least amount of steps as possible):**

1. Provision a Rancher v2.5 instance with a custom downstream cluster (I used https://github.com/superseb/tf-do-rancher2).
2. View the (empty) Ingresses view in the Cluster Explorer UI for the custom cluster, observe page loads as expected
3. Create some ingresses in the cluster:

```
for i in {1..100}; do
  cat << EOF | kubectl --kubeconfig=kubeconfig.yml apply -f -
  apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    name: test-$i
    namespace: default
  spec:
    rules:
    - host: test-2.example.cm
      http:
        paths:
        - backend:
            serviceName: kubernetes
            servicePort: 444
EOF
done
```
4. View the Ingresses view in the Cluster Explorer UI for the custom cluster again, now with Ingress resources present

**Result:**

Observe the page is slow to load and 100s of requests are generated to the Rancher proxied v1 endpoints for apps.relicasets, replicationcontrollers, batch.jobs, apps.daemonsets, batch.cronjobs, apps.deployments, apps.statefulsets (e.g. https://128.199.41.199/k8s/clusters/c-v5kqf/v1/batch.cronjobs).

**Other details that may be helpful:**

In the observed user environment, the user had c. 1500 deployments and c. 2000 replicasets. The traffic of the 100s of requests for the workload endpoints DoS's the cluster-agent/the controlplane in the downstream, ballooning cluster-agent memory until it is OOM killed, and the host temporarily freezes

**Environment information**
- Reproduced with Rancher v2.5.5 / v1.17.16-rancher1-1, matching user's initial environment, and Rancher v2.5.7 / v1.20.5-rancher1-1 to test latest

**Comment by paynejacob** [comment]
This seems to stem from how the rancher api implements pagination. We always request all records from the database and then only process the ones relevant for our page.  This reduces the payload size and minimizes how many records rancher needs to process but it still puts excessive load on the database.  To resolve this we need to implement pagination in our database queries.

https://github.com/rancher/norman/blob/c7fd1e24145bdb45af317615fd1c660001e456b6/api/handler/query.go#L55

![image](https://user-images.githubusercontent.com/12258347/125536543-c01b036f-c4da-4018-9e15-fb6e5956ff52.png)

**Comment by paynejacob** [comment]
After talking with @dramich Norman does not talk directly to the database, the kube api sits in between and supports pagination.  This will greatly simplify the solution.

**Comment by snasovich** [comment]
@paynejacob , per our discussion it looks like Cluster Explorer sorts by name by default, so the limitation you listed may prevent the suggested fix.
Another point is that new UI (Cluster Manager) uses /v1 API and doesn't pass any sorting or even limit. So filtering is likely done on the client. Is /v1 API also affected by this issue?

**Comment by paynejacob** [comment]
related https://github.com/rancher/rancher/pull/33283

**Comment by paynejacob** [comment]
# Explanation
​
## What is Norman?
​
Norman is an API used by the Rancher UI for interacting with downstream cluster's Kubernetes APIs.  It allows for a richer feature set than what is offered by the Kubernetes API. Operations like sorting, filtering, and transformations are possible with the Norman API.
​
## What happens in a Norman list request?
​
Norman requests can be broken into three main parts: deserialization, transformation and serialization. When a request is made to Norman the user's request is parsed and all objects of that resource type are requested from the Kubernetes API. All objects are listed because sorting is not always supported by the Kubernetes API. The results from the Kubernetes API are returned as json and parsed into an array of maps of strings to interfaces.  These maps are then transformed based on the user request or any additional transformations configured for the given endpoint.  This data is then serialized to json and sent to the requestor.
​
## How does Norman compare to kubectl?
​
When investigating the performance of listing resources in Norman kubectl was used as a baseline.  Kubectl was selected as a baseline because it lists an arbitrary resource in a cluster like Norman, however it does little to no transformation on the resources.  Kubectl was asked to get all resources of a given type in json format.  The json formatting was important because plain `get` in kubectl uses a special content type that returns only a few fields of a resource in a simplified table format from the Kubernetes API.  This allows kubectl to bypass the dynamic client and have near constant runtime time complexity.  Requesting json format forces kubectl to load the entire object and use the dynamic client making it a more accurate comparison to what Rancher does.  Using the unix time program the time it takes for kubectl to get a given resources was compared to the time it takes to curl a list of resources from Norman.  It was found that listing a resource in Norman with only the built in transformers is comparable to listing resources with kubectl.  These test were done using a k3d cluster and Rancher running on the test host.  This meant that any transfers would be made over the local network interface alleviating any concerns of the connection between Rancher and the downstream cluster adding significant overhead. The network overhead in real world situations can vary wildly and was not considered as part of this test.
​
## Why can listing resources with Norman be slow?

This section is a very technical deep dive into how responses from the Kubernetes API are deserialized.  The takeaway here is Norman uses a dynamic client to support any and all resources.  They dynamic client iteratively deserializes the response by iterating and copying the response body multiple times.  While this makes Norman very dynamic comes at the cost of preformance and it more noticable when listing large resources, such as configmaps or secrets.  
​
After an exhaustive deep dive into the performance of Norman endpoints it was determined most of the processing time comes from deserializing the Kubernetes API response and serializing the response.  Norman uses the Kubernetes Golang client with the `application/json` content type to interface with a given cluster.  There are two ways the client can interface with a cluster: using the `clientset` or using the REST API with the dynamic client. Norman uses the dynamic client to support any resource in any cluster without being compiled with the specific go structs. The `clientset` requires defined go structs for every resource type and is very performant when deserialzing data from the Kubernetes API. The dynamic client allows any arbitrary resource to be deseralized even if the application did not have the related struct at compile time. This comes at the cost of deseralizing requests taking 3 times as long as the `clientset`. This extra cost comes from how the dynamic client works.  When a response comes back from the Kubernetes API the response body bytes are read in their entirety into memory.  The response body is then normalized to json, in Norman's case this is a copy to a new array.  The response bytes are then unmarshalled into json.  This first unmarshall extracts a few top-level fields to inform the client what type of response this is.  In the case of a list operation the bytes are unmarshalled again to extract all the list metadata and the items. Finally, the response is unmarshalled a third time to unmarshal the list items.  Unmarshalling the body this many times adds up when there are large responses such as configmaps and secrets.  Using the dynamic client every byte in the response is iterated over and copied 4 times.  This can also consume a lot of memory as the response body is copied at least once and they each time the json is unmarshalled it is copied again.  After the response is deserialized it is passed through the various sorts, filters and transformers attached to the resource with Norman.  While the performance of these transforms can vary wildly based on how they are implemented poor performance here is usually isolated to one resource and is out of the control of Norman.  After the transformations are complete the response is rendered to json to and sent to the client.  On large responses this can be expensive because the json encoder must once again iterate over every byte.  With the number of copies listing resources with large response bodies, e.g. secrets with helm versions, can quickly fill up the memory on smaller clusters creating a resource constraint for other resources listed.  Given the Rancher UI lists most resources on startup this can result in poor performance across all endpoints making the UI take long time to load.
​
## What has been improved?
​
Norman stores now optionally accept a typer.  If a typer is given with the list type of the resource associated with the store registered, Norman can use this type to inform the go client of how to parse the response.  This eliminates two of the json unmarshales in the deserialization step greatly reducing memory and computation needed.  Additionally the `_drop` option was added.  This allows the API's client to exclude any top-level object field from the response reducing memory consumption and the computation required to serialize the response.  In cases where network overhead is a concern this can have the added benefit of smaller response bodies. We're going to be working with Rancher UI team to leverage this additional option.

For a testing environment I used a k3d cluster with rancher running locally.  All requests were done with curl piping the output to `/dev/null`. To test this I generated 200 configmaps with 8KB of data each.  I then listed them with the `application/json` content type three times and averaged the response time.  With no changes it took an average of `13,341 ms` to list configmaps.  Applying these changes but not registering `ConfigMapList` with the typer the response time averaged `12,719 ms`.  Registering the `ConfigMapList` with the typer brought it down to `7,764 ms`.  Adding `_drop=data` to the request query took the response time to `4,336ms`.

**Comment by vincent99** [comment]
Was any of this done for looking at the Steve API?  The large majority of requests in the current (explorer) UI never touch Norman, especially the ones that are frequently large responses, and the ones in the initial description of this issue...

**Comment by snasovich** [comment]
@vincent99 , the issue is reported against 2.5 and Cluster Explorer so I would assume it's v3 API (Norman). We will be looking into v1 API (Steve) in a separate issue.

**Comment by vincent99** [comment]
No.  Explorer is the new UI and uses almost exclusively v1/steve (especially in 2.5.x with no multi-cluster management stuff).  Every request shown here from users is to Steve.  Only you guys are talking about Norman.

This actual issue appears to be a UI bug somehow getting into a (possibly eventually terminating?) loop loading the 7 workload types to find things related to the ingress.  It should never be loading the same type repeatedly.

API performance is at best incidental here, in that it's apparently easily possible to DOS the API server with requests from one browser.  But not related to the problem described.

**Comment by paynejacob** [comment]
@vincent99 the explanation here is only for Norman but after investigating the same issue applies to steve.  and the solution will be similar,  There are multiple performance issues and the work for them is being done all at once, that explanation should be on another issue but the general reasoning above will apply to steve as well.  If there is a ui bug that causes infinite fetches that will trigger the performance bug in the API as well.  Fixing the ui bug will make the API issue manifest less often but it will still manifest for users with many resources.

**Comment by vincent99** [comment]
If you go to the ingress page, without having previously loaded the workload types from some other screen, each row independently makes 7 requests to the [ui] store for all those types so it can find the info for the Target column.

Each of those look at the store's state, sees that it doesn't have that type loaded yet, and makes an API request to load them.  We would normally recognize this, only make one actual API call and have all the others wait for the same answer, but [this](https://github.com/rancher/dashboard/blob/21fc3058df808fe4937acdb0092bafd844908899/plugins/steve/actions.js#L36) top-notch implementation of that I wrote 2 years ago doesn't work for some reason

>     // @TODO queue/defer duplicate requests

So the net result is 7 * len(ingresses) XmlHttpRequests given to the browser at roughly the same time.  With HTTP/1.1 the client would generally limit that to 4-6 actual requests [per domain] in flight at the same time.  But with HTTP/2 that's up to the server, and often basically unlimited (looks like 128 by default for nginx).

I'll fix the UI, but you probably still want to find the ingress annotations to lower that to prevent self-owns.

**Comment by vincent99** [comment]
Also I would be more than happy to change the client-side (for Steve, at least) to take responses that are something like newline-delimited-JSON, or individual frames sent over the websockets instead of one giant JSON object string.  Parsing them isn't fast on the client-side either, and blocks UI interaction, and both parse & stringily likely scale more-than-linearly with the object size.

**Comment by ibuildthecloud** [comment]
This odd, why do we care at all about Norman? In Steve I think we specifically drop helm secret data because it's huge.

**Comment by paynejacob** [comment]
The research on this started during 2.5 and the other issue was specifically for the ember ui.  I took over this issue because steve sees similar problems and the fix for norman could be applied to steve as well.

**Comment by junkiebev** [comment]
May I ask if we could limit this happening as often by requesting that our users stop using the Cluster Explorer and stick with legacy Cluster Manager until a fix is shipped? We have a few cases open about it, and I'd like to collapse them into one with a temporary remediation.

**Comment by vincent99** [comment]
The particular issue in the beginning of this thread is a "tidal wave" of duplicate requests specifically from loading the ingress list page.  You can certainly ask people not to use that page (or to go to the workload list first, which causes them to already be loaded and not trigger the tidal wave).  Or it will be fixed in 2.6.1 soon (but is not backported into 2.5.x).

**Comment by junkiebev** [comment]
Oh does this not have anything to do with cluster agents disconnecting on timeouts?￼ Separate issue?￼ 

Engineering thought the two may be related - ref

https://github.com/rancher/rancher/issues/34819

**Comment by paynejacob** [comment]
### Root cause
The dashboard was generating duplicate requests for listing resources
 
### What was fixed, or what changes have occurred
The dashboard should only list resources once per page navigation
 
### Areas or cases that should be tested
listing ingresses, configmaps and secrets
 
### What areas could experience regressions?
Any list page
 
### Are the repro steps accurate/minimal?
yes

**Comment by sgapanovich** [comment]
2 setups: 2.5-head and 2.5.11
both have 250 ingresses deployed 
- there are no improvements for ingress page in dashboard:
  - still ~700 requests 
  - chrome page froze on me a few times and the tab became not responsive 
![same.jpg](https://zube.io/files/rancher/b9297e3e56c1e0269a509881699ed8d1-same.jpg)

cc: @paynejacob @vincent99

**Comment by sgapanovich** [comment]
I have two setups: 2.5-head and 2.5.11
Both have ~250 ingresses 
I don't see any improvements for the ingress page in dashboard:
- ~700 requests are made to get resources 
- chrome tab froze when i tried to go though pages 
![same](https://user-images.githubusercontent.com/40282141/141873601-128c3023-cb2a-4adc-8299-c644e94fefa2.jpg)

cc: @paynejacob @vincent99

**Comment by vincent99** [comment]
The UI change is only in 2.6.

**Comment by sgapanovich** [comment]
Issue is re-opened
reason: original issue reported is not fixed

**Comment by snasovich** [comment]
@SheilaghM , this issue targets 2.5.13 and has a complicated history as several somewhat related issues got bundled here.
The main fix (prevention of additional requests from UI to backend as explained here https://github.com/rancher/rancher/issues/30718#issuecomment-925194541) is fixed in 2.6.1+ (https://github.com/rancher/dashboard/issues/4637), but the lift to backport it to 2.5.x is quite large, especially considering the fact that Vue UI is not the main UI there.

So, I'm recommending having this particular issue closed and communicating that performance improvements to new (Vue) UI are available in 2.6.1+.

**Comment by deniseschannon** [comment]
It looks as if this is and will only be fixed as part of 2.6 line due to the comments. 

@sgapanovich Can you verify that this is the case for 2.6-head?

**Comment by sgapanovich** [comment]
@deniseschannon I verified that a few days ago here 
https://github.com/rancher/dashboard/issues/4637#issuecomment-982046146
